### DISCUSSION TITLE
cargo-mutants 23.12.0

### DISCUSSION THREAD

**Discussion by sourcefrog** [sourcefrog/cargo-mutants#181] [Announcements]
An exciting step forward: cargo-mutants can now generate mutations smaller than a whole function. To start with, several binary operators are mutated.

Also, `--in-diff` will no longer error out if no mutants are generated in the diff, making it much more reasonable for a CI check.

- New: Mutate `==` to `!=` and vice versa.

- New: Mutate `&&` to `||` and vice versa, and mutate both of them to `==` and `!=`.

- New: Mutate `<`, `<=`, `>`, `>=`.

- Changed: If no mutants are generated then `cargo mutants` now exits successfully, showing a warning. (Previously it would exit with an error.) This works better with `--in-diff` in CI, where it's normal that some changes may not have any mutants.

- Changed: Include column numbers in text listings of mutants and output to disambiguate smaller-than-function mutants, for example if there are several operators that can be changed on one line. This also applies to the names used for regex matching, so may break some regexps that match the entire line (sorry). The new option `--line-col=false` turns them both off in `--list` output.

- Changed: In the mutants.json format, replaced the `function`, `line`, and `return_type` fields with a `function` submessage (including the name and return type) and a `span` indicating the entire replaced region, to better handle smaller-than-function mutants. Also, the `function` includes the line-column span of the entire function.

<hr /><em>This discussion was created from the release <a href='https://github.com/sourcefrog/cargo-mutants/releases/tag/v23.12.0'>cargo-mutants 23.12.0</a>.</em>

**Comment by dmoka** [comment]
Appreciate you work you do to make this tool more and more powerful! 💪
